### PR TITLE
Avoids sparse array in StateSet and State

### PR DESCRIPTION
**Pull request by cedricpinson**
It does not seems to have a big effect on performance sample even if it seems a bit faster on chrome and similar in firefox, but on safari it seems more interesting

this the timing of 'render' in stats
chrome:
3.6/3.7 (master)
3.3/3.45 (avoid-sparse-array)

firefox:
4.7/4.9 (master)
4.5/4.8 (avoid-sparse-array)

safari:
6.7/6.9 (master)
5.5/5.8 (avoid-sparse-array)

Also adds a better handling of Scissor

### CODE REVIEW

File: sources/osg/State.js [hunk 1]
```diff
@@ -480,11 +490,18 @@ MACROUTILS.createPrototypeObject(
...
                 for (var j = 0; j < lt; j++) {
                     var attributeId = j;
-                    var attributeStack = _textureAttributeArray[attributeId];
-                    var stateSetAttributePair = stateSetTextureAttributeArray
-                        ? stateSetTextureAttributeArray[attributeId]
-                        : undefined;
-                    var hasStateAttributeStack = attributeStack !== undefined;
+
+                    var attributeStack =
+                        attributeId < _textureAttributeArray.length
```
Comment (Kuranes) [inline sources/osg/State.js]: tested caching _textureAttributeArray.length and stateSetTextureAttributeArray.length as var outside of the loop ?

File: sources/osg/State.js [hunk 2]
```diff
@@ -803,58 +822,65 @@ MACROUTILS.createPrototypeObject(
...
 
         pushAttributeMap: function(_attributeArray, stateSetAttributeArray, validAttributeArray) {
             /*jshint bitwise: false */
             var attributeStack;
-
             for (var i = 0, l = validAttributeArray.length; i < l; i++) {
                 var index = validAttributeArray[i];
-                var attributePair = stateSetAttributeArray[index];
+                var attributePair =
+                    index < stateSetAttributeArray.length
```
Comment (Kuranes) [inline sources/osg/State.js]: ( same note about caching length )

File: sources/osgShader/ShaderGenerator.js [hunk 3]
```diff
@@ -100,7 +101,7 @@ ShaderGenerator.prototype = {
 
             for (var i = 0; i < cacheType.length; i++) {
                 var type = cacheType[i];
-                var attributeStack = textureUnit[type];
+                var attributeStack = type < textureUnit.length ? textureUnit[type] : undefined;
```
Comment (Kuranes) [inline sources/osgShader/ShaderGenerator.js]: (same note about textureUnit.length and _attributeArrayForUnit.length )

Comment (cedricpinson) [reply sources/osgShader/ShaderGenerator.js]: fixed

File: sources/osg/State.js (hunk 2 above)
Comment (cedricpinson) [reply sources/osg/State.js]: fixed

File: sources/osg/State.js (hunk 1 above)
Comment (cedricpinson) [reply sources/osg/State.js]: fixed

**Pull request merged**